vault/blog.2022.2022-04-05.growing-knowledge-bases-from-the-bottom-up.md

```diff
@@ -0,0 +1,174 @@
+
+Aha! We commonly implement search algorithms using tree data structures. Using the bottom-up structure of tags, we’ve pieced together knowledge that we can use this to expand our Python hierarchy. We already have a place in our hierarchy for data-structures, so we can add a note under this hierarchy as a type of data-structure–a **tree**. 
+
+After creating the note `lang.python.data.tree`, we may even decide that we have obvious groupings, like Binary Search Trees, AVL Trees, and B-trees, and create sub-notes for these. 
+
+![algo to hierarchy](https://org-dendron-public-assets.s3.amazonaws.com/images/bottom-2.png)
+
+Tags are a common bottom-up structure in knowledge bases that allow us to build distributed groups. If we navigate to the note for `#tree`, we find backlinks to the tagged chapters arranged in a group. 
+
+![backlinks](https://org-dendron-public-assets.s3.amazonaws.com/images/bottom-3.png)
+
+As we move to integrate these tags into our knowledge base, we return to the problem at hand: we want to implement a search algorithm in Python. We realize the binary search tree is exactly what we need. We create a note with the implementation that we end up using.
+
+```markdown
+<!-- lang.python.data.tree -->
+
+## Summary
+
+An example of a ![[tags.tree]]
+
+## Example
+class Node:
+    def init(self, val): 
+        self.l_child = None 
+        self.r_child = None 
+        self.data = val 
+
+def binary_insert(root, node): 
+    if root is None: 
+        root = node
+    else: 
+         if root.data > node.data:
+              if root.l_child is None: 
+                  root.l_child = node 
+              else: 
+                 binary_insert(root.l_child, node) 
+         else: 
+             if root.r_child is None: 
+                root.r_child = node 
+             else: 
+                binary_insert(root.r_child, node)
+
+- Example from [here](stackoverflow.com/questions/5444394/how-to-implement-a-binary...)
+```
+
+The [note reference](https://wiki.dendron.so/notes/f1af56bb-db27-47ae-8406-61a98de6c78c/) `![[tags.tree]]` will show embed the contents of the `#tree` note, rendering as: 
+
+```markdown
+## Summary
+
+Trees are ...
+```
+
+This allows us to connect our specific Python implementation with the general knowledge we are accumulating about trees. If we navigate to the `tag.tree` note, we also can access everywhere the tag is used in our knowledge base through [backlinks](https://pkm.dendron.so/notes/0486be25-19e4-4fc6-98f6-e5c9908d1161/).
+
+With tags, we can work up from our informational units, forming clusters of content as we go. We form groups, then we define categories. In doing so, we extend our top-down structure using bottom-up structures. 
+
+Peanut Butter, meet Jelly.
```

Comment:
at the very end of the article, can you add the following? this reminds folks to subscribe to the blog

```
---

![[blog.subscribe]]
```